### PR TITLE
rule_validation added

### PR DESCRIPTION
**Pull request by MaciejPoland**
Validation of rules added when Player is putting a card from hand to table.
Validation of rules added when first card is putted on table.

### CODE REVIEW

**Review by MateuszKarolChrusciel — CHANGES_REQUESTED** [review]
please correct as discussed

File: macao.py [hunk 1]
```diff
@@ -26,11 +34,19 @@ def __init__(self, num_of_players):
             player.draw_initial_hand(self.deck)
 
         # lay out the first card on the table
-        self.table.append(self.deck.pool.pop(random.randint(0, len(self.deck.pool) - 1)))
+        correct_table = False
+        while correct_table == False:
+            if len(self.table) == 0:
+                self.table.append(self.deck.pool.pop(random.randint(0, len(self.deck.pool) - 1)))
```
Comment (MateuszKarolChrusciel) [inline macao.py]: pls correct

File: macao.py [hunk 2]
```diff
@@ -26,11 +34,19 @@ def __init__(self, num_of_players):
             player.draw_initial_hand(self.deck)
 
         # lay out the first card on the table
-        self.table.append(self.deck.pool.pop(random.randint(0, len(self.deck.pool) - 1)))
+        correct_table = False
+        while correct_table == False:
```
Comment (MateuszKarolChrusciel) [inline macao.py]: while correct_table == False: -> while not correct_table:

File: macao.py [hunk 3]
```diff
@@ -26,11 +34,19 @@ def __init__(self, num_of_players):
             player.draw_initial_hand(self.deck)
 
         # lay out the first card on the table
-        self.table.append(self.deck.pool.pop(random.randint(0, len(self.deck.pool) - 1)))
```
Comment (MateuszKarolChrusciel) [inline macao.py:29]: move this to be executed after checking for correct card

File: macao.py [hunk 4]
```diff
@@ -26,11 +34,19 @@ def __init__(self, num_of_players):
...
+                self.table.append(self.deck.pool.pop(random.randint(0, len(self.deck.pool) - 1)))
+            if not self.table[len(self.table) - 1].rank in ("Ace", "King", "Queen", "Jack", "Four", "Three", "Two"):
+                correct_table = True
+                self.table.append(self.deck.pool.pop(random.randint(0, len(self.deck.pool) - 1)))
 
         # start the game
         self._main()
 
+
+
```
Comment (MateuszKarolChrusciel) [inline macao.py]: try ctrl+alt+L

File: macao.py [hunk 5]
```diff
@@ -8,6 +8,14 @@ class Deck:
...
         self.pool = cards.generate_all_cards()
 
+def validate_card(table_card, player_card):
+    return any(
+        (table_card.rank == "Queen",
+         player_card.rank == "Queen",
+         table_card.rank == player_card.rank,
+         table_card.suit == player_card.suit,
+         )
+    )
```
Comment (MateuszKarolChrusciel) [inline macao.py]: move under classes

File: macao.py [hunk 6]
```diff
@@ -51,8 +67,9 @@ def _main(self):
                     print(f"{card_index}. {repr(card)}")
                 # set flag "users_selection_is_valid" to False
                 users_selection_is_valid = False
+                validation_rule = False
```
Comment (MateuszKarolChrusciel) [inline macao.py]: change to "is_compliant_with_rules"

**Review by MateuszKarolChrusciel — CHANGES_REQUESTED** [review]

File: macao.py (hunk 1 above)
Comment (MaciejPoland) [reply macao.py]: corrected

File: macao.py (hunk 2 above)
Comment (MaciejPoland) [reply macao.py]: corrected

File: macao.py (hunk 3 above)
Comment (MaciejPoland) [reply macao.py:29]: corrected; later, once is_active will be added to a Class card can be changed to be alligned to class atribute

File: macao.py (hunk 4 above)
Comment (MaciejPoland) [reply macao.py]: corrected

File: macao.py (hunk 5 above)
Comment (MaciejPoland) [reply macao.py]: corrected

File: macao.py (hunk 6 above)
Comment (MaciejPoland) [reply macao.py]: corrected

File: macao.py [hunk 7]
```diff
@@ -26,7 +26,15 @@ def __init__(self, num_of_players):
             player.draw_initial_hand(self.deck)
 
         # lay out the first card on the table
-        self.table.append(self.deck.pool.pop(random.randint(0, len(self.deck.pool) - 1)))
+        # self.table.append(self.deck.pool.pop(random.randint(0, len(self.deck.pool) - 1)))
+        correct_table = False
+        while not correct_table:
+            drawn_card = self.deck.pool.pop(random.randint(0, len(self.deck.pool) - 1))
+            if drawn_card in ("Ace", "King", "Queen", "Jack", "Four", "Three", "Two"):
+                self.deck.append(card_to_be_drawn)
```
Comment (MateuszKarolChrusciel) [inline macao.py]: please change to "self.deck.pool.append(drawn_card)"

or i.e. "card_to_be_laid_out" if u wish

**Review by MateuszKarolChrusciel — APPROVED** [review]
IS OK, YAY

**Review by SlawekMaciejewski — APPROVED** [review]
OK

**Pull request merged**